app/Main.hs

```diff
@@ -0,0 +1,66 @@
+{-# LANGUAGE OverloadedStrings #-}
+
+module Main where
+
+import           Cmd
+import           Control.Monad.Trans.Class
+import           Data.Aeson                (object, (.=))
+import           Data.Document
+import           Data.Text
+import           Data.Time
+import           Database.MongoDB          hiding (Document, runCommand)
+import           Network.HTTP.Types.Status
+import           Options
+import           Web.Scotty                hiding (Options)
+
+-------------------------------------------------------------------------------------------------------------------------
+-- Mongo initialisation & teardown
+-------------------------------------------------------------------------------------------------------------------------
+data Mongo = Mongo {
+  mongoCon        :: Pipe,
+  mongoDatabase   :: Text,
+  mongoCollection :: Text
+  }
+
+mongoConnect :: MongoOpts -> IO Mongo
+mongoConnect (MongoOpts hst db col) =  do
+  pipe <- connect hst
+  _    <- access pipe master db $ ensureIndex dateIndex
+  return $ Mongo pipe db col
+    where dateIndex = Index col ["date" =: (1 :: Int)] "creation_date" False False (Just 86400)
+
+-------------------------------------------------------------------------------------------------------------------------
+-- Web service
+-------------------------------------------------------------------------------------------------------------------------
+receiveAd :: Mongo -> ActionM ()
+receiveAd (Mongo pipe db col) = do
+  domain  <- param "domain"
+  time    <- lift getCurrentTime
+  content <- jsonData
+  _       <- lift $ access pipe master db $ insertDocument col (Document time domain content)
+  status status201
+
+listAds :: Mongo -> ActionM ()
+listAds (Mongo pipe db col) = do
```

Comment:
Instead of pattern matching here, I'd use functions to access whatever I need from the `Mongo`. If you add a type to the record, then it will require editing every single pattern match. But if you use functions, you don't need to worry about that.

Reply:
Should be fixed in 24e7bde2bcc268d64f3c32bbab00daa4b263af27.